cpp/src/io/parquet/reader_impl_preprocess.cu

```diff
@@ -1259,6 +1259,90 @@ struct start_offset_output_iterator {
   }
 };
 
+struct flat_column_num_rows {
+  gpu::PageInfo const* pages;
+  gpu::ColumnChunkDesc const* chunks;
+
+  __device__ size_type operator()(size_type pindex)
+  {
+    gpu::PageInfo const& page = pages[pindex];
+    // ignore dictionary pages and pages belonging to any column containing repetition (lists)
+    if ((page.flags & gpu::PAGEINFO_FLAGS_DICTIONARY) ||
+        (chunks[page.chunk_idx].max_level[gpu::level_type::REPETITION] > 0)) {
+      return 0;
+    }
+    return page.num_rows;
+  }
+};
+
+struct row_counts_nonzero {
+  __device__ bool operator()(size_type count) { return count > 0; }
+};
+
+struct row_counts_different {
+  size_type const expected;
+  __device__ bool operator()(size_type count) { return (count != 0) && (count != expected); }
+};
+
+/**
+ * @brief Detects malformed parquet input data.
+ *
+ * We have seen cases where parquet files can be oddly malformed. This function specifically
+ * detects one case in particular:
+ *
+ * - When when you have a file containing N rows
+ * - For some reason, the sum total of the number of rows over all pages for a given column
+ *   is != N
+ *
+ * @param pages All pages to be decoded
+ * @param chunks Chunk data
+ * @param page_keys Keys (schema id) associated with each page, sorted by column
+ * @param page_index Page indices for iteration, sorted by column
+ * @param expected_row_count Expected row count, if applicable
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ */
+void detect_malformed_pages(hostdevice_vector<gpu::PageInfo>& pages,
+                            hostdevice_vector<gpu::ColumnChunkDesc> const& chunks,
+                            device_span<const int> page_keys,
+                            device_span<const int> page_index,
+                            std::optional<size_t> expected_row_count,
+                            rmm::cuda_stream_view stream)
+{
+  // sum row counts for all non-dictionary, non-list columns. other columns will indicated as 0
+  rmm::device_uvector<int> row_counts(pages.size(), stream);  // worst case:  num keys == num pages
+  auto size_iter = thrust::make_transform_iterator(
+    page_index.begin(), flat_column_num_rows{pages.device_ptr(), chunks.device_ptr()});
+  auto row_counts_begin                     = row_counts.begin();
+  [[maybe_unused]] auto [_, row_counts_end] = thrust::reduce_by_key(rmm::exec_policy(stream),
+                                                                    page_keys.begin(),
+                                                                    page_keys.end(),
+                                                                    size_iter,
+                                                                    thrust::make_discard_iterator(),
+                                                                    row_counts_begin);
+
+  auto nonzero = thrust::find_if(
+    rmm::exec_policy(stream), row_counts_begin, row_counts_end, row_counts_nonzero{});
```

Comment:
`find_if` is known to have performance regression. Use `thrust::count_if(...) > 0` instead.

Reply:
One option is to do a `copy_if` (of nonzero values) to a new buffer and use the result of that to check the condition.  I avoided that initially because I have a pathological fear of memory allocation.  But maybe with the pool allocator it's no big deal.

Also worth pointing out:  the size of the input here is at the very most in the thousands.  Typically it will be much less.  So the overhead of `find_if` may not be so bad.

Reply:
Actually, what I can do is just double the size of the row_counts allocation and copy_if to the second half of the vector.  Although that may be a little too cowboy for you guys :)

Reply:
My inclination based on the size is that it's not going to matter a whole lot to the overall performance of this function which option we select. @ttnghia do you have a better recollection of exactly how bad the slowdown from `find_if` could be, and whether it's relevant at these data sizes? Without a strong performance argument I'm pretty indifferent to which solution we pick. I suspect that any performance gains/losses in this check (whether from an extra allocation, more kernel launches, or a slow algorithm) will probably be washed out by the rest of the overall parquet reader logic given the small size of the inputs to this function.

Reply:
`thrust::find_if` is the building block of logical operators `thrust::any_of`/`thrust::any_of` etc. 
The issue is also a long time issue: https://github.com/NVIDIA/thrust/issues/1016.

Reply:
Yes, I'm familiar with that issue, but the data sizes there are massive (billions of elements) whereas we are talking about thousands. As a result I suspect it won't actually matter here.